File: ainow/services/model_manager.py
"""
Model manager — starts/stops llama-server with different GGUF models.

Singleton that manages the llama-server process lifecycle.
Started once at server boot via CLI --model flag.

Models are configured via environment variables:
  LLAMA_SERVER_EXE  — path to llama-server binary
  MODELS_DIR        — base directory for GGUF model files
  MODEL_<ALIAS>     — override path for a specific model (e.g. MODEL_GEMMA=/path/to/model.gguf)
"""

import os
import time
import subprocess
from typing import Optional

import httpx

from ..log import get_logger

logger = get_logger("ainow.model_manager")

LLAMA_SERVER_EXE = os.getenv("LLAMA_SERVER_EXE", "llama-server")
LLAMA_SERVER_PORT = int(os.getenv("LLAMA_SERVER_PORT", "8080"))
MODELS_DIR = os.getenv("MODELS_DIR", os.path.join(os.path.expanduser("~"), "models"))

# Short aliases for CLI (--model 4b)
MODEL_ALIASES = {
    "0.8b": "qwen3.5-0.8b",
    "2b": "qwen3.5-2b",
    "4b": "qwen3.5-4b",
    "9b": "qwen3.5-9b",
    "27b": "qwen3.5-27b",
    "online": "openrouter-gemini-flash-lite",
}


def _model_path(*parts):
    """Build a model path relative to MODELS_DIR."""
    return os.path.join(MODELS_DIR, *parts)


def _build_models():
    """Build model configs. Env var MODEL_<ALIAS> overrides the default path."""
    models = {
        "qwen3.5-0.8b": {
            "name": "Qwen 0.8B",
            "hf_repo": "lmstudio-community/Qwen3.5-0.8B-GGUF",
            "hf_files": ["Qwen3.5-0.8B-Q8_0.gguf", "mmproj-Qwen3.5-0.8B-BF16.gguf"],
            "model": _model_path("Qwen3.5-0.8B-GGUF", "Qwen3.5-0.8B-Q8_0.gguf"),
            "mmproj": _model_path("Qwen3.5-0.8B-GGUF", "mmproj-Qwen3.5-0.8B-BF16.gguf"),
        },
        "qwen3.5-2b": {
            "name": "Qwen 2B",
            "hf_repo": "lmstudio-community/Qwen3.5-2B-GGUF",
            "hf_files": ["Qwen3.5-2B-Q4_K_M.gguf", "mmproj-Qwen3.5-2B-BF16.gguf"],
            "model": _model_path("Qwen3.5-2B-GGUF", "Qwen3.5-2B-Q4_K_M.gguf"),
            "mmproj": _model_path("Qwen3.5-2B-GGUF", "mmproj-Qwen3.5-2B-BF16.gguf"),
        },
        "qwen3.5-4b": {
            "name": "Qwen 4B",
            "hf_repo": "lmstudio-community/Qwen3.5-4B-GGUF",
            "hf_files": ["Qwen3.5-4B-Q4_K_M.gguf", "mmproj-Qwen3.5-4B-BF16.gguf"],
            "model": _model_path("Qwen3.5-4B-GGUF", "Qwen3.5-4B-Q4_K_M.gguf"),
            "mmproj": _model_path("Qwen3.5-4B-GGUF", "mmproj-Qwen3.5-4B-BF16.gguf"),
        },
        "qwen3.5-9b": {
            "name": "Qwen 9B",
            "hf_repo": "lmstudio-community/Qwen3.5-9B-GGUF",
            "hf_files": ["Qwen3.5-9B-UD-Q4_K_XL.gguf", "mmproj-Qwen3.5-9B-BF16.gguf"],
            "model": _model_path("Qwen3.5-9B-GGUF", "Qwen3.5-9B-UD-Q4_K_XL.gguf"),
            "mmproj": _model_path("Qwen3.5-9B-GGUF", "mmproj-Qwen3.5-9B-BF16.gguf"),
        },
        "qwen3.5-27b": {
            "name": "Qwen 27B",
            "hf_repo": "lmstudio-community/Qwen3.5-27B-GGUF",
            "hf_files": ["Qwen3.5-27B-UD-IQ3_XXS.gguf", "mmproj-BF16.gguf"],
            "model": _model_path("Qwen3.5-27B-GGUF", "Qwen3.5-27B-UD-IQ3_XXS.gguf"),
            "mmproj": _model_path("Qwen3.5-27B-GGUF", "mmproj-BF16.gguf"),
            "ctx": "32768",
        },
        "openrouter-gemini-flash-lite": {
            "name": "Gemini Flash Lite (OpenRouter)",
            "online": True,
            "base_url": "https://openrouter.ai/api/v1",
            "api_key_env": "OPENROUTER_API_KEY",
            "model_id": "google/gemini-3.1-flash-lite-preview",
        },
    }

    # Register extra models from env vars: MODEL_GEMMA=/path/to/dir
    # Format: MODEL_<ALIAS>=<model_gguf_path>;<mmproj_gguf_path>;<display_name>
    # Or:     MODEL_<ALIAS>=<directory>  (auto-detect .gguf and mmproj files)
    for key, value in os.environ.items():
        if not key.startswith("MODEL_") or key in ("MODELS_DIR", "MODEL_ALIASES"):
            continue
        alias = key[6:].lower()  # MODEL_GEMMA -> gemma
        if alias in MODEL_ALIASES or alias in models:
            continue

        if ";" in value:
            # Explicit: MODEL_GEMMA=/path/model.gguf;/path/mmproj.gguf;Gemma 4B
            parts = value.split(";")
            model_file = parts[0].strip()
            mmproj_file = parts[1].strip() if len(parts) > 1 else ""
            display_name = parts[2].strip() if len(parts) > 2 else alias.upper()
        elif os.path.isdir(value):
            # Directory: auto-detect model and mmproj files
            model_file = ""
            mmproj_file = ""
            display_name = alias.upper()
            for f in os.listdir(value):
                fl = f.lower()
                if fl.endswith(".gguf"):
                    if "mmproj" in fl:
                        mmproj_file = os.path.join(value, f)
                    elif not model_file:
                        model_file = os.path.join(value, f)
        else:
            # Single file path — assume it's the model, no mmproj
            model_file = value
            mmproj_file = ""
            display_name = alias.upper()

        if model_file:
            MODEL_ALIASES[alias] = alias
            config = {"name": display_name, "model": model_file}
            if mmproj_file:
                config["mmproj"] = mmproj_file
            models[alias] = config
            logger.info(f"Registered custom model '{alias}' from env")

    return models


MODELS = _build_models()

COMMON_ARGS = [
    "-c", "262144",
    "-np", "1",
    "--fit", "on",
    "--fit-target", "1024",
    "-fa", "on",
    "-t", "20",
    "--no-mmap",
    "--jinja",
    "-ctk", "q4_0",
    "-ctv", "q4_0",
    "--reasoning", "off",
    "--reasoning-budget", "0",
    "--port", str(LLAMA_SERVER_PORT),
]


def _ensure_llama_server():
    """Download llama-server from GitHub if not found."""
    global LLAMA_SERVER_EXE

    if os.path.isfile(LLAMA_SERVER_EXE):
        return

    # Only auto-download if using default name (not a custom path)
    if os.path.dirname(LLAMA_SERVER_EXE):
        # User specified a full path that doesn't exist
        raise FileNotFoundError(f"llama-server not found at: {LLAMA_SERVER_EXE}")

    import platform
    import zipfile

    logger.info("llama-server not found, downloading latest release...")

    # Detect platform
    system = platform.system().lower()  # windows, linux, darwin

    # Get latest release info
    resp = httpx.get(
        "https://api.github.com/repos/ggml-org/llama.cpp/releases/latest",
        timeout=30,
    )
    resp.raise_for_status()
    release = resp.json()
    tag = release["tag_name"]

    # Find the right asset
    target = None
    cudart_target = None
    for asset in release["assets"]:
        name = asset["name"]
        if system == "windows" and "win" in name and "cuda" in name and name.endswith(".zip"):
            if "cudart" not in name:
                target = asset
            else:
                cudart_target = asset
        elif system == "linux" and "linux" in name and "cuda" in name and name.endswith(".zip"):
            if "cudart" not in name:
                target = asset
            else:
                cudart_target = asset

    if not target:
        raise RuntimeError(f"No suitable llama-server release found for {system}")

    # Download and extract next to MODELS_DIR
    install_dir = os.path.join(os.path.dirname(MODELS_DIR), "llama-server")
    os.makedirs(install_dir, exist_ok=True)

    for asset in [target, cudart_target]:
        if not asset:
            continue
        url = asset["browser_download_url"]
        zip_path = os.path.join(install_dir, asset["name"])

        logger.info(f"Downloading {asset['name']}...")
        with httpx.stream("GET", url, follow_redirects=True, timeout=300) as r:
            r.raise_for_status()
            with open(zip_path, "wb") as f:
                for chunk in r.iter_bytes(8192):
                    f.write(chunk)

        with zipfile.ZipFile(zip_path, "r") as z:
            z.extractall(install_dir)
        os.remove(zip_path)

    # Update the global to point to the downloaded binary
    exe_name = "llama-server.exe" if system == "windows" else "llama-server"
    LLAMA_SERVER_EXE = os.path.join(install_dir, exe_name)

    if not os.path.isfile(LLAMA_SERVER_EXE):
        raise RuntimeError(f"Download succeeded but {exe_name} not found in {install_dir}")

    logger.info(f"llama-server {tag} installed to {install_dir}")


def _ensure_model_files(config):
    """Download model files from HuggingFace if not present."""
    if "hf_repo" not in config:
        return  # Custom or online model, user manages files

    from huggingface_hub import hf_hub_download

    for filename in config.get("hf_files", []):
        # The subfolder in MODELS_DIR is derived from the HF repo name's last part
        subfolder = config["hf_repo"].split("/")[-1]
        local_path = os.path.join(MODELS_DIR, subfolder, filename)

        if os.path.exists(local_path):
            continue

        logger.info(f"Downloading {filename} from {config['hf_repo']}...")
        os.makedirs(os.path.dirname(local_path), exist_ok=True)
        hf_hub_download(
            repo_id=config["hf_repo"],
            filename=filename,
            local_dir=os.path.join(MODELS_DIR, subfolder),
        )
        logger.info(f"Downloaded {filename}")


def resolve_model_id(name: str) -> str:
    """Resolve a short alias (e.g. '4b') or full id to a full model id."""
    if name in MODEL_ALIASES:
        return MODEL_ALIASES[name]
    if name in MODELS:
        return name
    raise ValueError(f"Unknown model: {name}. Valid: {', '.join(MODEL_ALIASES.keys())}")


class ModelManager:
    """Singleton managing the llama-server process."""

    def __init__(self):
        self._process: Optional[subprocess.Popen] = None
        self._current_model: Optional[str] = None

    @property
    def current_model(self) -> Optional[str]:
        return self._current_model

    def _check_health_sync(self) -> bool:
        """Synchronous health check."""
        try:
            r = httpx.get(
                f"http://localhost:{LLAMA_SERVER_PORT}/health",
                timeout=2.0,
            )
            return r.status_code == 200
        except Exception:
            return False

    def _stop_server_sync(self) -> None:
        """Stop the current llama-server process (synchronous)."""
        if self._process and self._process.poll() is None:
            try:
                self._process.terminate()
                time.sleep(2)
                if self._process.poll() is None:
                    self._process.kill()
                    time.sleep(0.5)
            except Exception as e:
                logger.error(f"Error stopping server: {e}")
            self._process = None
        else:
            # Try to kill whatever is on the port
            try:
                result = subprocess.run(
                    ["netstat", "-ano"],
                    stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                    text=True, encoding="utf-8", errors="replace", timeout=5,
                )
                for line in result.stdout.splitlines():
                    if f":{LLAMA_SERVER_PORT}" in line and "LISTENING" in line:
                        parts = line.split()
                        pid = int(parts[-1])
                        subprocess.run(
                            ["taskkill", "/PID", str(pid), "/F"],
                            capture_output=True, timeout=5,
                        )
                        logger.info(f"Killed existing process on port {LLAMA_SERVER_PORT} (PID {pid})")
                        time.sleep(1)
                        break
            except Exception as e:
                logger.error(f"Error killing existing process: {e}")

        self._current_model = None

    def start(self, model_id: str) -> None:
        """
        Start llama-server with the given model (synchronous, blocking).

        Called once at server boot before the event loop starts.
        Kills any existing llama-server first.
        """
        if model_id not in MODELS:
            raise ValueError(f"Unknown model: {model_id}")

        config = MODELS[model_id]

        # Auto-download llama-server and model files if missing
        if not config.get("online"):
            _ensure_llama_server()
            _ensure_model_files(config)

        logger.info(f"Starting llama-server with {config['name']}...")

        # Stop any existing server
        if self._check_health_sync():
            logger.info("Stopping existing llama-server...")
            self._stop_server_sync()

        # Start new server
        cmd = [LLAMA_SERVER_EXE, "-m", config["model"]]
        if config.get("mmproj"):
            cmd += ["--mmproj", config["mmproj"]]
        cmd += COMMON_ARGS
        # Per-model context override
        if "ctx" in config:
            cmd[cmd.index("-c") + 1] = config["ctx"]

        self._process = subprocess.Popen(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            creationflags=subprocess.CREATE_NEW_PROCESS_GROUP,
        )

        # Wait for health check
        max_wait = 120
        poll_interval = 1.0
        elapsed = 0.0

        while elapsed < max_wait:
            time.sleep(poll_interval)
            elapsed += poll_interval

            if self._process.poll() is not None:
                raise RuntimeError(f"llama-server exited unexpectedly (code {self._process.returncode})")

            if self._check_health_sync():
                self._current_model = model_id
                logger.info(f"llama-server ready with {config['name']} ({elapsed:.0f}s)")
                return

            if int(elapsed) % 10 == 0:
                logger.info(f"  Waiting for llama-server... ({elapsed:.0f}s)")

        raise TimeoutError(f"llama-server did not become healthy within {max_wait}s")

    def stop(self) -> None:
        """Stop llama-server (synchronous)."""
        self._stop_server_sync()


# Singleton
model_manager = ModelManager()

File: src-tauri/src/model_manager.rs
//! Model manager -- starts/stops llama-server with different GGUF models.
//!
//! Manages the llama-server process lifecycle:
//! - Start/stop llama-server as a child process
//! - Health check polling
//! - Auto-download from GitHub releases (llama-server binary)
//! - Auto-download models from HuggingFace
//! - Kill existing process on port

use anyhow::{bail, Result};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use tokio::process::{Child, Command};

const LLAMA_SERVER_PORT: u16 = 8080;

/// Configuration for a model.
#[derive(Debug, Clone)]
pub struct ModelConfig {
    pub name: String,
    pub model_path: Option<String>,
    pub mmproj_path: Option<String>,
    pub hf_repo: Option<String>,
    pub hf_files: Vec<String>,
    pub ctx: Option<String>,
    /// If true, this is an online model (no llama-server needed).
    pub online: bool,
    pub base_url: Option<String>,
    pub api_key_env: Option<String>,
    pub model_id: Option<String>,
}

/// Common llama-server arguments.
const COMMON_ARGS: &[&str] = &[
    "-c", "262144",
    "-np", "1",
    "--fit", "on",
    "--fit-target", "1024",
    "-fa", "on",
    "-t", "20",
    "--no-mmap",
    "--jinja",
    "-ctk", "q4_0",
    "-ctv", "q4_0",
    "--reasoning", "off",
    "--reasoning-budget", "0",
];

/// Build the default model registry.
pub fn build_models(models_dir: &str) -> HashMap<String, ModelConfig> {
    let md = |subdir: &str, file: &str| -> String {
        Path::new(models_dir).join(subdir).join(file).to_string_lossy().to_string()
    };

    let mut models = HashMap::new();

    models.insert("qwen3.5-0.8b".to_string(), ModelConfig {
        name: "Qwen 0.8B".to_string(),
        model_path: Some(md("Qwen3.5-0.8B-GGUF", "Qwen3.5-0.8B-Q8_0.gguf")),
        mmproj_path: Some(md("Qwen3.5-0.8B-GGUF", "mmproj-Qwen3.5-0.8B-BF16.gguf")),
        hf_repo: Some("lmstudio-community/Qwen3.5-0.8B-GGUF".to_string()),
        hf_files: vec!["Qwen3.5-0.8B-Q8_0.gguf".to_string(), "mmproj-Qwen3.5-0.8B-BF16.gguf".to_string()],
        ctx: None, online: false, base_url: None, api_key_env: None, model_id: None,
    });

    models.insert("qwen3.5-4b".to_string(), ModelConfig {
        name: "Qwen 4B".to_string(),
        model_path: Some(md("Qwen3.5-4B-GGUF", "Qwen3.5-4B-Q4_K_M.gguf")),
        mmproj_path: Some(md("Qwen3.5-4B-GGUF", "mmproj-Qwen3.5-4B-BF16.gguf")),
        hf_repo: Some("lmstudio-community/Qwen3.5-4B-GGUF".to_string()),
        hf_files: vec!["Qwen3.5-4B-Q4_K_M.gguf".to_string(), "mmproj-Qwen3.5-4B-BF16.gguf".to_string()],
        ctx: None, online: false, base_url: None, api_key_env: None, model_id: None,
    });

    models.insert("qwen3.5-9b".to_string(), ModelConfig {
        name: "Qwen 9B".to_string(),
        model_path: Some(md("Qwen3.5-9B-GGUF", "Qwen3.5-9B-UD-Q4_K_XL.gguf")),
        mmproj_path: Some(md("Qwen3.5-9B-GGUF", "mmproj-Qwen3.5-9B-BF16.gguf")),
        hf_repo: Some("lmstudio-community/Qwen3.5-9B-GGUF".to_string()),
        hf_files: vec!["Qwen3.5-9B-UD-Q4_K_XL.gguf".to_string(), "mmproj-Qwen3.5-9B-BF16.gguf".to_string()],
        ctx: None, online: false, base_url: None, api_key_env: None, model_id: None,
    });

    models.insert("qwen3.5-27b".to_string(), ModelConfig {
        name: "Qwen 27B".to_string(),
        model_path: Some(md("Qwen3.5-27B-GGUF", "Qwen3.5-27B-UD-IQ3_XXS.gguf")),
        mmproj_path: Some(md("Qwen3.5-27B-GGUF", "mmproj-BF16.gguf")),
        hf_repo: Some("lmstudio-community/Qwen3.5-27B-GGUF".to_string()),
        hf_files: vec!["Qwen3.5-27B-UD-IQ3_XXS.gguf".to_string(), "mmproj-BF16.gguf".to_string()],
        ctx: Some("32768".to_string()),
        online: false, base_url: None, api_key_env: None, model_id: None,
    });

    models.insert("online".to_string(), ModelConfig {
        name: "Online".to_string(),
        model_path: None, mmproj_path: None,
        hf_repo: None, hf_files: vec![],
        ctx: None, online: true,
        base_url: Some("https://openrouter.ai/api/v1".to_string()),
        api_key_env: Some("ONLINE_API_KEY".to_string()),
        model_id: Some("google/gemini-3.1-flash-lite-preview".to_string()),
    });

    models
}

/// Short aliases for model IDs.
pub fn model_aliases() -> HashMap<String, String> {
    let mut m = HashMap::new();
    m.insert("0.8b".to_string(), "qwen3.5-0.8b".to_string());
    m.insert("4b".to_string(), "qwen3.5-4b".to_string());
    m.insert("9b".to_string(), "qwen3.5-9b".to_string());
    m.insert("27b".to_string(), "qwen3.5-27b".to_string());
    m.insert("online".to_string(), "online".to_string());
    m
}

/// Manages the llama-server process.
pub struct ModelManager {
    process: Option<Child>,
    current_model: Option<String>,
    llama_server_exe: String,
    models_dir: String,
    port: u16,
}

impl ModelManager {
    pub fn new(llama_server_exe: String, models_dir: String) -> Self {
        Self {
            process: None,
            current_model: None,
            llama_server_exe,
            models_dir,
            port: LLAMA_SERVER_PORT,
        }
    }

    pub fn current_model(&self) -> Option<&str> {
        self.current_model.as_deref()
    }

    /// Check if llama-server is healthy.
    async fn check_health(&self) -> bool {
        let url = format!("http://localhost:{}/health", self.port);
        match reqwest::Client::new()
            .get(&url)
            .timeout(std::time::Duration::from_secs(2))
            .send()
            .await
        {
            Ok(resp) => resp.status().is_success(),
            Err(_) => false,
        }
    }

    /// Stop the current llama-server process.
    pub async fn stop(&mut self) {
        if let Some(mut proc) = self.process.take() {
            let _ = proc.kill().await;
            let _ = proc.wait().await;
            log::info!("llama-server stopped");
        } else {
            // Try to kill whatever is on the port (Windows)
            self.kill_port_process().await;
        }
        self.current_model = None;
    }

    /// Kill any process listening on our port (Windows).
    async fn kill_port_process(&self) {
        if cfg!(windows) {
            let output = Command::new("cmd")
                .args(["/C", &format!("netstat -ano | findstr :{} | findstr LISTENING", self.port)])
                .stdout(Stdio::piped())
                .stderr(Stdio::null())
                .output()
                .await;

            if let Ok(output) = output {
                let text = String::from_utf8_lossy(&output.stdout);
                for line in text.lines() {
                    let parts: Vec<&str> = line.split_whitespace().collect();
                    if let Some(pid_str) = parts.last() {
                        if let Ok(pid) = pid_str.parse::<u32>() {
                            let _ = Command::new("taskkill")
                                .args(["/PID", &pid.to_string(), "/F"])
                                .stdout(Stdio::null())
                                .stderr(Stdio::null())
                                .output()
                                .await;
                            log::info!("Killed existing process on port {} (PID {})", self.port, pid);
                            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
                        }
                    }
                }
            }
        }
    }

    /// Ensure the llama-server binary exists, downloading if necessary.
    async fn ensure_llama_server(&mut self) -> Result<()> {
        let exe_path = Path::new(&self.llama_server_exe);

        if exe_path.is_file() {
            return Ok(());
        }

        // Only auto-download if using default name (not a custom path)
        if exe_path.parent().is_some() && exe_path.parent() != Some(Path::new("")) {
            bail!("llama-server not found at: {}", self.llama_server_exe);
        }

        log::info!("llama-server not found, downloading latest release...");

        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(30))
            .build()?;

        let resp = client
            .get("https://api.github.com/repos/ggml-org/llama.cpp/releases/latest")
            .header("User-Agent", "AINow")
            .send()
            .await?;

        let release: serde_json::Value = resp.json().await?;
        let tag = release["tag_name"].as_str().unwrap_or("unknown");

        let system = if cfg!(windows) {
            "win"
        } else if cfg!(target_os = "linux") {
            "linux"
        } else {
            "darwin"
        };

        let assets = release["assets"].as_array().ok_or_else(|| anyhow::anyhow!("No assets in release"))?;

        let mut target = None;
        let mut cudart_target = None;
        for asset in assets {
            let name = asset["name"].as_str().unwrap_or("");
            if name.contains(system) && name.contains("cuda") && name.ends_with(".zip") {
                if !name.contains("cudart") {
                    target = Some(asset);
                } else {
                    cudart_target = Some(asset);
                }
            }
        }

        let target = target.ok_or_else(|| anyhow::anyhow!("No suitable llama-server release found for {system}"))?;

        let install_dir = Path::new(&self.models_dir)
            .parent()
            .unwrap_or(Path::new("."))
            .join("llama-server");
        tokio::fs::create_dir_all(&install_dir).await?;

        for asset in [Some(target), cudart_target].into_iter().flatten() {
            let url = asset["browser_download_url"].as_str().unwrap_or("");
            let name = asset["name"].as_str().unwrap_or("download.zip");
            let zip_path = install_dir.join(name);

            log::info!("Downloading {name}...");
            let resp = reqwest::Client::new()
                .get(url)
                .send()
                .await?;
            let bytes = resp.bytes().await?;
            tokio::fs::write(&zip_path, &bytes).await?;

            // Extract zip using system unzip or PowerShell
            if cfg!(windows) {
                let _ = Command::new("powershell")
                    .args([
                        "-Command",
                        &format!(
                            "Expand-Archive -Path '{}' -DestinationPath '{}' -Force",
                            zip_path.display(),
                            install_dir.display()
                        ),
                    ])
                    .output()
                    .await?;
            } else {
                let _ = Command::new("unzip")
                    .args(["-o", &zip_path.to_string_lossy(), "-d", &install_dir.to_string_lossy()])
                    .output()
                    .await?;
            }
            tokio::fs::remove_file(&zip_path).await?;
        }

        let exe_name = if cfg!(windows) {
            "llama-server.exe"
        } else {
            "llama-server"
        };
        let installed_path = install_dir.join(exe_name);
        if !installed_path.is_file() {
            bail!("Download succeeded but {exe_name} not found in {}", install_dir.display());
        }

        self.llama_server_exe = installed_path.to_string_lossy().to_string();
        log::info!("llama-server {tag} installed to {}", install_dir.display());
        Ok(())
    }

    /// Ensure model files exist, downloading from HuggingFace if necessary.
    async fn ensure_model_files(&self, config: &ModelConfig) -> Result<()> {
        let hf_repo = match &config.hf_repo {
            Some(r) => r,
            None => return Ok(()),
        };

        let subfolder = hf_repo.split('/').last().unwrap_or(hf_repo);
        let local_dir = Path::new(&self.models_dir).join(subfolder);

        for filename in &config.hf_files {
            let local_path = local_dir.join(filename);
            if local_path.exists() {
                continue;
            }

            log::info!("Downloading {filename} from {hf_repo}...");
            tokio::fs::create_dir_all(&local_dir).await?;

            let url = format!(
                "https://huggingface.co/{}/resolve/main/{}",
                hf_repo, filename
            );

            let client = reqwest::Client::builder()
                .redirect(reqwest::redirect::Policy::limited(10))
                .build()?;
            let resp = client.get(&url).send().await?;

            if !resp.status().is_success() {
                bail!("Failed to download {filename}: {}", resp.status());
            }

            let bytes = resp.bytes().await?;
            tokio::fs::write(&local_path, &bytes).await?;
            log::info!("Downloaded {filename}");
        }

        Ok(())
    }

    /// Start llama-server with the given model.
    pub async fn start(&mut self, model_id: &str) -> Result<()> {
        let models = build_models(&self.models_dir);
        let config = models
            .get(model_id)
            .ok_or_else(|| anyhow::anyhow!("Unknown model: {model_id}"))?
            .clone();

        if config.online {
            log::info!("Online model selected, no llama-server needed");
            self.current_model = Some(model_id.to_string());
            return Ok(());
        }

        // Auto-download llama-server and model files if missing
        self.ensure_llama_server().await?;
        self.ensure_model_files(&config).await?;

        log::info!("Starting llama-server with {}...", config.name);

        // Stop any existing server
        if self.check_health().await {
            log::info!("Stopping existing llama-server...");
            self.stop().await;
        }

        // Build command
        let model_path = config.model_path.as_deref().unwrap_or("");
        let mut args: Vec<String> = vec!["-m".to_string(), model_path.to_string()];
        if let Some(mmproj) = &config.mmproj_path {
            args.push("--mmproj".to_string());
            args.push(mmproj.clone());
        }
        for arg in COMMON_ARGS {
            args.push(arg.to_string());
        }
        args.push("--port".to_string());
        args.push(self.port.to_string());

        // Per-model context override
        if let Some(ctx) = &config.ctx {
            if let Some(pos) = args.iter().position(|a| a == "-c") {
                if pos + 1 < args.len() {
                    args[pos + 1] = ctx.clone();
                }
            }
        }

        let child = Command::new(&self.llama_server_exe)
            .args(&args)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()?;

        self.process = Some(child);

        // Wait for health check
        let max_wait = 120;
        let poll_interval = std::time::Duration::from_secs(1);
        let mut elapsed = 0u64;

        while elapsed < max_wait {
            tokio::time::sleep(poll_interval).await;
            elapsed += 1;

            // Check if process exited
            if let Some(ref mut proc) = self.process {
                if let Ok(Some(status)) = proc.try_wait() {
                    bail!("llama-server exited unexpectedly (code {:?})", status.code());
                }
            }

            if self.check_health().await {
                self.current_model = Some(model_id.to_string());
                log::info!("llama-server ready with {} ({elapsed}s)", config.name);
                return Ok(());
            }

            if elapsed % 10 == 0 {
                log::info!("  Waiting for llama-server... ({elapsed}s)");
            }
        }

        bail!("llama-server did not become healthy within {max_wait}s");
    }
}
